plugins/elanfp/fu-elanfp-usb-device.c

```diff
@@ -0,0 +1,557 @@
+					    error)) {
+		g_prefix_error(error, "failed to release interface: ");
+		return FALSE;
+	}
+
+	/* FuUsbDevice->close */
+	return FU_DEVICE_CLASS(fu_elanfp_usb_device_parent_class)->close(device, error);
+}
+
+static gboolean
+fu_elanfp_usb_device_do_xfer(FuElanfpUsbDevice *self,
+			     guint8 *outbuf,
+			     gsize outlen,
+			     guint8 *inbuf,
+			     gsize inlen,
+			     gboolean allow_less,
+			     gsize *rxed_count,
+			     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer(usb_device,
+						ELAN_EP_CMD_OUT,
+						outbuf,
+						outlen,
+						&actual,
+						BULK_SEND_TIMEOUT_MS,
+						NULL,
+						error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_PARTIAL_INPUT,
+				    "only sent %" G_GSIZE_FORMAT "/%" G_GSIZE_FORMAT " bytes",
+				    actual,
+				    outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer(usb_device,
+						ELAN_EP_IMG_IN,
+						inbuf,
+						inlen,
+						&actual,
+						BULK_RECV_TIMEOUT_MS,
+						NULL,
+						error)) {
+			return FALSE;
+		}
+		if (actual != inlen && !allow_less) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_PARTIAL_INPUT,
+				    "only received %" G_GSIZE_FORMAT "/%" G_GSIZE_FORMAT " bytes",
+				    actual,
+				    outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count != NULL)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_elanfp_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuElanfpUsbDevice *self = FU_ELANFP_USB_DEVICE(device);
+	const gchar *fw_ver_str = NULL;
+	guint16 fw_ver;
+	guint8 usb_buf[2] = {0x40, 0x19};
+	gsize actual_len = 0;
+
+	if (!fu_elanfp_usb_device_do_xfer(self,
+					  (guint8 *)&usb_buf,
+					  sizeof(usb_buf),
+					  usb_buf,
+					  sizeof(usb_buf),
+					  TRUE,
+					  &actual_len,
+					  error)) {
+		g_prefix_error(error, "failed to device setup: ");
+		return FALSE;
+	}
+
+	fw_ver = fu_common_read_uint16(usb_buf, G_BIG_ENDIAN);
+
+	fw_ver_str = g_strdup_printf("%04x", fw_ver);
+
+	g_debug("fw version %04x", fw_ver);
+
+	fu_device_set_version(device, fw_ver_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_elanfp_usb_device_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	FuElanfpUsbDevice *self = FU_ELANFP_USB_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	if (!run_iap_process(self, fw, error)) {
+		g_prefix_error(error, "device write firmware - iap fail: ");
+		return FALSE;
+	}
+
+	g_debug("device write firmware - iap success !!");
```

Comment:
this debugging line can probably go given you have error handling plumbed through